tempesta_fw/http_parser.c

```diff
@@ -103,17 +92,104 @@ memchreol(const unsigned char *s, size_t n)
 #define IS_WS(c) (c == ' ' || c == '\t')
 
 /**
+ * When parsing it's necessary to not forget to update @data and @len
+ * properly. So, __data_{} macros help to reduce the amount of direct
+ * @data/@len manipulations.
+ */
+#define __data_offset(pos)						\
+	(size_t)((pos) - data)
+#define __data_remain(pos)						\
+	(len - __data_offset(pos))
+#define __data_rebase(pos)						\
+	len -= __data_offset(pos), data = pos;
+#define __data_available(pos, num)					\
+	(num <= __data_remain(pos))
+
+/**
+ * Every header field which is parsed must be handled via the following set of
+ * macroses. Being once opened with @__msg_hdr_open, the header may be updated
+ * with @__msg_hdr_fixup or finished with @__msg_hdr_finish. The latter must be
+ * used on EOL only.
+ */
+#define __msg_hdr_open(pos)						\
+do {									\
+	tfw_http_msg_hdr_open(msg, pos);				\
+	__data_rebase(pos);						\
+} while (0)
+#define __msg_hdr_fixup(pos)						\
+do {									\
+	if (__data_offset(pos)) {					\
+		tfw_http_msg_hdr_chunk_fixup(msg, data,			\
+					     __data_offset(pos));	\
+		__data_rebase(pos);					\
+	}								\
+} while (0)
+#define __msg_hdr_finish(pos)						\
+({									\
+	int eolen = 1 + !!((parser->_tmp.eol & 0xff) == 0xda);		\
+	tfw_str_set_eolen(&parser->hdr, eolen);				\
+	tfw_http_msg_hdr_close(msg, msg->parser._hdr_tag);		\
+})
```

Comment:
The macros are relatively complex, so please use empty lines between them (they are almost functions).